File: source/yeon/10주차/code/1300k번째수.cpp
// 백준 1300 k 번째수
// cpp solved by binary search

#include<iostream>
#include<algorithm>
int N, k;

int main(void){
    scanf("%d\n%d", &N,&k);
    int left = 1, right = k, ans = 0;
    while(left <= right){
        int mid = (left + right)/2;

        int cnt = 0;
        for(int i = 1; i < N + 1; i++){
            cnt += std::min(mid/i, N);
        }

        if (cnt >= k){
            ans = mid;
            right = mid - 1;
        }else{
            left = mid + 1;
        }
    }
    
    printf("%d", ans);
    return 0;
}

File: source/yeon/10주차/code/1300k번째수.py
# 백준 1300 k번째 수 
# python solved by binary search

N, k = int(input()), int(input())
left, right = 1, k

# parametric Search 
# check mid is k'th element or not 
# count the number of element under mid 
# and binary search according to compare result of cnt & k
while left <= right:
    # just set mid as half of arr
    # so time complexity can be log(k)
    mid = (left + right)//2

    cnt = 0
    # count the number of element under mid 
    for i in range(1,N+1):
        cnt += min(mid//i, N)

    # and binary search according to compare result of cnt & k
    if cnt >= k:
        ans = mid
        right = mid -1
    else:
        left = mid + 1

print(ans)